### PyTorch

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/c/c6/PyTorch_logo_black.svg/1920px-PyTorch_logo_black.svg.png"
     width="300" style="background:white">

- 공식 사이트 : https://pytorch.org/
- 한글 공식 사이트 : https://pytorch.kr/
- 공식 사이트의 튜토리얼 한글화 사이트 : https://tutorials.pytorch.kr/

#### PyTorch 설치
- CPU 설치
    - 간단하지만, 그래픽 가속 사용 불가
    - 기본 파이썬 환경 설치

- GPU 설치
    - 가상환경 생성, 설치

##### CPU설치
- https://pytorch.org/get-started/locally/ 에 나와있는 Note에서 해당 플랫폼 종류 선택, 명령어 복사 실행
```python
!pip install torch torchvision torchaudio
```

In [ ]:
!pip install torch torchvision torchaudio

In [3]:
import torch

In [4]:
## 2.4.0+cpu = CPU용 설치
torch.__version__

'2.4.0+cpu'

##### PyTorch GPU 설치
- NVIDIA GPU에서만 가능, CUDA로 되어있기 때문
- GPU 설치가 가능한지 먼저 확인

```python
!nvidia-smi
```

In [5]:
!nvidia-smi

Tue Aug  6 12:32:05 2024       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 536.23                 Driver Version: 536.23       CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                     TCC/WDDM  | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA GeForce GTX 1650      WDDM  | 00000000:01:00.0  On |                  N/A |
| 50%   33C    P8               8W /  75W |    786MiB /  4096MiB |     10%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+--

##### CUDA Version 12.2
- CUDA(Compute Unified Device Architecture) - NVIDIA에서 그래픽카드 가속을 위해 만든 기술
- https://ko.wikipedia.co.kr

##### CUDA 설치
- CUDA Toolkit 12.2
    - https://developer.nvidia.com/cuda-12-2-2-download-archive?target_os=Windows&target_arch=x86_64&target_version=11&target_type=exe_local
- cuDNN(Deep Natural Network)
    - CUDA Toolkit에 맞는 버전 확인
    - https://developer.nvidia.com/rdp/cudnn-archive 확인 후 다운로드(단, 로그인 필수)

#### GPU용 파이썬 가상환경 만들기
- GPU PyTorch는 설치 로그를 봐야하기 때문에 명령프롬프트에서 진행

- 터미널 오픈
    - virtualenv 확인
    - pip install virtualenv
    - virtualenv gpu_env
- 가상환경 활성화
    - cd gpu_env\Scripts\activate

- PyTorch GPU 설치
    - PyTorch GetStarted 사이트에서 확인 후 명령어
    - pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu121

In [1]:
import torch

In [2]:
## 2.4.0+cu121 --> CUDA 12.1 버전의 PyTorch
torch.__version__

'2.4.0+cu121'

In [3]:
## CUDA 사용가능 여부 확인
torch.cuda.is_available()

True

In [4]:
## 설치된 CUDA가 동작하는 그래픽카드 정보확인
torch.cuda.get_device_name(0)

'NVIDIA GeForce GTX 1650'

##### 이제부터 GPU 가속으로 PyTorch(텐서플로 가능) 딥러닝을 할 수 있음